### PR TITLE
layout fixes and compute hold

### PR DESCRIPTION
**Pull request by sepiroth887**
n/a

**Summary of Changes**

More accurate hold page display including Computed Hold. Display and functionality as per Honeywell ref. 
Distance calculation is based off TAS prediction at WP using a rough estimate, still missing accurate GS prediction using wind at WP

Changelog:

* redesign of hold page to closer match ref spec
* automatically computed hold based on waypoint profile (alt, ref speed)

**Screenshots (if necessary)**

https://youtu.be/PcDyFecQHsI

**References**

Honeywell R1A Jan 2009
Restrictions: 
image

GUI:
image

image

image

**Additional context**

Hold execution logic has been started but mostly commented out and needs a lot of work before it can be enabled. FS9GPS is likely the only path here. 

Also, ref material has Hold at waypoint ident colored green, but there seems to be no way right now to change the color of only part of the string. Any attempts to split into columns would break the entire view. Not 100% sure this is my fault or just a drawback on Asobo's mcdu display handling right now. 

Discord username (if different from GitHub):

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js [hunk 1]
```diff
@@ -296,6 +296,34 @@ class CDUFlightPlanPage {
                                 }
                             };
                         }
+
+                        if (mcdu.activeHold && mcdu.activeHold.has(waypoint.ident)) {
+                            let holdRows = [
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:301]: `const` not `let`.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js [hunk 2]
```diff
@@ -296,6 +296,34 @@ class CDUFlightPlanPage {
...
+                            ];
+
+                            // place the button input on the HOLD line to clear the hold
+                            mcdu.onRightInput[i+1] = async () => {
+                                // TODO remove any active waypoints and route to active waypoint
+                                mcdu.activeHold = null
+                                CDUFlightPlanPage.ShowPage(mcdu, offset);
+                            }
+                            
+                            for(let j = 0; i < rowsCount-1; i++) {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:316]:
```suggestion
                            for (let j = 0; i < rowsCount-1; i++) {
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js [hunk 3]
```diff
@@ -296,6 +296,34 @@ class CDUFlightPlanPage {
...
+                                rows[2 * i + 1] = holdRows[j++];
+                                if (i == rowsCount-2 || j >= holdRows.length) {
+                                    break
+                                }
+                                
+                                rows[2 * i + 2] = holdRows[j++]; 
+                            }
+                        }else {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:324]:
```suggestion
                        } else {
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js [hunk 4]
```diff
@@ -1053,6 +1054,91 @@ class FlightPlanManager {
         SimVar.SetSimVarValue("C:fs9gps:FlightPlanNewApproachTransition", "number", transition);
         SimVar.SetSimVarValue("C:fs9gps:FlightPlanLoadApproach", "number", 1);
     }
+    setApproach(approach)  {
+        this._approach = approach;
+    }
+    calculateInitialPattern(turn, distance, course, waypoint) {
+        // console.log("constructing entry pattern for hold at "+ waypoint.ident + "")
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1061]: Remove commented code

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js [hunk 5]
```diff
@@ -1053,6 +1054,91 @@ class FlightPlanManager {
...
+        let courseRad = course * (Math.PI / 180.0)
+        let parallelEntry = [courseRad, ((course + 110) % 360) * (Math.PI / 180.0)];
+        let tearEntry = [courseRad, ((course - 70) % 360) * (Math.PI / 180.0)]
+        let directEntry = [((course+110) % 360) * (Math.PI / 180.0), ((course + 290) % 360) * (Math.PI / 180.0)]
+
+        // let planeHeading = Simplane.getHeadingMagnetic();
+        let headingRad = planeHeading * (Math.PI / 180.0);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1068]: Those should be `const`

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js [hunk 6]
```diff
@@ -1053,6 +1054,91 @@ class FlightPlanManager {
...
+        }
+        if (headingRad > Math.min(directEntry[0], directEntry[1]) && headingRad < Math.max(directEntry[0], directEntry[1])) {
+            entry = "DIRECT"
+        }
+
+        let waypoints = [new LatLong(waypoint.infos.lat, waypoint.infos.long)] // start with the hold fix
+
+        let reverseTurnBearing = turn == "L" ? (course + 90) % 360 : (course - 90) % 360;
+        let turnBearing = turn == "R" ? (course + 45) % 360 : (course - 90) % 360;
+        let backCourse = (course + 180) %360
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1086]: Those should be `const`

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js [hunk 7]
```diff
@@ -1053,6 +1054,91 @@ class FlightPlanManager {
...
+        //         waypoints.push(lastWP)
+        //         // return to fix on hold track
+        //         lastWP = Avionics.Utils.bearingDistanceToCoordinates(course, distance, waypoints[0].lat, waypoints[0].long);
+        //         break;
+        //     case "DIRECT":
+        //         // nothing to do here
+        //         break;
+        // }
+
+        // return waypoints;
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1115]: Can you clarify why this code is commented ?

If your PR is not ready yet, please make it draft 🙂

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js [hunk 8]
```diff
@@ -1053,6 +1054,91 @@ class FlightPlanManager {
...
+        }catch (err) {        
+            console.log(err)
+        }
+    }
+    // isHolding() {
+    //     return true;
+    // }
+    // updateHoldPath() {
+    //     console.log("tick")
+    // }
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1141]: Ditto

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_HoldAtPage.js [hunk 9]
```diff
@@ -51,56 +51,251 @@ class CDUHoldAtPage {
...
+                                mcdu.flightPlanManager.setHoldApproach(mcdu.holdTurn, mcdu.distComp, mcdu.holdCourse, waypoint);
+                                mcdu.manualHoldData = null
+                                CDUFlightPlanPage.ShowPage(mcdu);
+                            };
+                        }catch(err){
+                            console.log(err)
+                        }
+                    }
                 }
+                //# sourceMappingURL=A320_Neo_CDU_HoldAtPage.js.map
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_HoldAtPage.js:231]:
```suggestion

```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_HoldAtPage.js [hunk 10]
```diff
@@ -51,56 +51,251 @@ class CDUHoldAtPage {
         mcdu.onLeftInput[0] = () => {
                 let value = mcdu.inOut;
                 if (isNaN(value) || 0 < value > 360) {
-                    mcdu.inOut = "NaN"
+                    mcdu.inOut = "NaN";
                     return;
+                }class CDUHoldAtPage {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_HoldAtPage.js:56]:
```suggestion
                }
                
                class CDUHoldAtPage {
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_HoldAtPage.js [hunk 11]
```diff
@@ -51,56 +51,251 @@ class CDUHoldAtPage {
...
+                    static ShowPage(mcdu, waypoint, waypointIndexFP) {
+                        try {
+                            let SpeedConstraints = {
```
Comment (Meqolo) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_HoldAtPage.js:59]: Should be a const, as this isnt changed

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js (hunk 7 above)
Comment (sepiroth887) [reply A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1115]: Mentioned on the pr description. Let me get this commented out code onto another branch for now.

File: A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js (hunk 6 above)
Comment (sepiroth887) [reply A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1086]: Are const arrays still growable/changeable?

Wasn't aware const even exists 👍

Comment (wuperguy) [reply A32NX/html_ui/Pages/VCockpit/Instruments/NavSystems/LogicElements/A32NX_FlightPlanManager.js:1086]: Const just means the variable cannot be reassigned. Stands for "constant", you can still mutate the array and call various functions upon it.

**Comment by NathanInnes** [comment]
Please include a changelog.

**Comment by sepiroth887** [comment]
> Please include a changelog.

Added in pr description

**Comment by NathanInnes** [comment]
I mean commit to the CHANGELOG.md located in .github > CHANGELOG.md :)

**Comment by AdenFlorian** [comment]
Please add screenshots of the relevant pages of the honeywell.

**Comment by sepiroth887** [comment]
Lol it was late @NathanInnes  sorry. Ill update that and add screenshots soon

**Comment by sepiroth887** [comment]
gah who enabled linters! :D root of all evil!  Fine i shall obey.

**Comment by AdenFlorian** [comment]
npm install
npm run lint-fix
commit
profit

**Comment by sepiroth887** [comment]
alright, a bit a rebase hell, merge master chaos, manual copy pasta from old change set ontop of new master and i think that's the one that should work :) 

lint applied, thanks for enabling it :)

**Comment by AdenFlorian** [comment]
@sepiroth887 you can set your PR to be a draft PR if you don't want us to merge it yet :)

**Comment by AdenFlorian** [comment]
tested and LGTM!

**Pull request merged**